File: src/main/java/com/sothawo/springdataelastictest/Person.java
/*
 * (c) Copyright 2019 sothawo
 */
package com.sothawo.springdataelastictest;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.domain.Persistable;
import org.springframework.data.elasticsearch.annotations.DateFormat;
import org.springframework.data.elasticsearch.annotations.Document;
import org.springframework.data.elasticsearch.annotations.Field;
import org.springframework.data.elasticsearch.annotations.FieldType;
import org.springframework.lang.Nullable;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

/**
 * @author P.J. Meisch (pj.meisch@sothawo.com)
 */
@Document(indexName = "person")
public class Person implements Persistable<Long> {

    @Id
    @Nullable
    private Long id;

    @Field(type = FieldType.Text, fielddata = true)
    @Nullable
    private String lastName;

    @Field(type = FieldType.Text, fielddata = true)
    @Nullable
    private String firstName;

    @Field(type = FieldType.Date, format = DateFormat.basic_date)
    @Nullable
    private LocalDate birthDate;

    @Field(type = FieldType.Nested)
    @Nullable
    private List<Movie> movies;

    @CreatedDate
    @Field(type = FieldType.Date, format = DateFormat.basic_date_time)
    @Nullable
    private Instant created;
    @CreatedBy
    @Nullable
    private String createdBy;
    @LastModifiedDate
    @Field(type = FieldType.Date, format = DateFormat.basic_date_time)
    @Nullable
    private Instant lastModified;
    @LastModifiedBy
    @Nullable
    private String lastModifiedBy;

    @Nullable
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    @Nullable
    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    @Nullable
    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    @Nullable
    public LocalDate getBirthDate() {
        return birthDate;
    }

    public void setBirthDate(LocalDate birthDate) {
        this.birthDate = birthDate;
    }

    @Nullable
    public Instant getCreated() {
        return created;
    }

    public void setCreated(Instant created) {
        this.created = created;
    }

    @Nullable
    public String getCreatedBy() {
        return createdBy;
    }

    public void setCreatedBy(String createdBy) {
        this.createdBy = createdBy;
    }

    @Nullable
    public Instant getLastModified() {
        return lastModified;
    }

    public void setLastModified(Instant lastModified) {
        this.lastModified = lastModified;
    }

    @Nullable
    public String getLastModifiedBy() {
        return lastModifiedBy;
    }

    public void setLastModifiedBy(String lastModifiedBy) {
        this.lastModifiedBy = lastModifiedBy;
    }

    @Nullable
    public List<Movie> getMovies() {
        return movies;
    }

    public void setMovies(@Nullable List<Movie> movies) {
        this.movies = movies;
    }

    @Override
    @JsonIgnore
    public boolean isNew() {
        return id == null || (createdBy == null && created == null);
    }

    @Override
    public String toString() {
        return "Person{" +
            "id=" + id +
            ", lastName='" + lastName + '\'' +
            ", firstName='" + firstName + '\'' +
            ", birthDate=" + birthDate +
            ", movies=" + movies +
            ", created=" + created +
            ", createdBy='" + createdBy + '\'' +
            ", lastModified=" + lastModified +
            ", lastModifiedBy='" + lastModifiedBy + '\'' +
            '}';
    }
}

File: src/main/kotlin/com/sothawo/springdataelastictest/population/Person.kt
package com.sothawo.springdataelastictest.population

import com.fasterxml.jackson.annotation.JsonIgnore
import com.github.javafaker.Faker
import org.springframework.data.annotation.CreatedBy
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.LastModifiedBy
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.annotation.ReadOnlyProperty
import org.springframework.data.annotation.Version
import org.springframework.data.domain.Persistable
import org.springframework.data.elasticsearch.annotations.DateFormat
import org.springframework.data.elasticsearch.annotations.Document
import org.springframework.data.elasticsearch.annotations.Field
import org.springframework.data.elasticsearch.annotations.FieldType
import org.springframework.data.elasticsearch.annotations.Mapping
import org.springframework.data.elasticsearch.annotations.ScriptedField
import org.springframework.lang.Nullable
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale

/**
 * @author P.J. Meisch (pj.meisch@sothawo.com)
 */
@Document(indexName = "population-person")
data class Person(

    @Id
    var id: String? = null,

    @Field(type = FieldType.Text)
    var lastName: String? = null,

    @Field(type = FieldType.Text)
    var firstName: String? = null,

    @Field(type = FieldType.Keyword)
    var house: String? = null
) {
    companion object {

        private val FAKER: Faker = Faker(Locale.GERMANY)

        fun create(id: String): Person = Person(
            id = id,
            firstName = FAKER.name().firstName(),
            lastName = FAKER.name().lastName(),
        )
    }
}
